cpp/dolfin/fem/DofMapPermuter.cpp

```diff
@@ -0,0 +1,524 @@
+    }
+
+    start += face_dofs;
+    for (std::size_t j = 0; j < face.size(); ++j)
+    {
+      rotation1[face[j]] = face[base_faces[0][j]];
+      rotation2[face2[j]] = face2[base_faces[0][j]];
+      reflection[face[j]] = face[base_faces[1][j]];
+    }
+  }
+
+  std::vector<int> rotation3(volume_dofs);
+  for (int j = 0; j < volume_dofs; ++j)
+    rotation3[j] = rotation2[rotation2[rotation1[j]]];
+  return {rotation1, rotation2, rotation3, reflection};
+}
+//-----------------------------------------------------------------------------
+Eigen::Array<std::int8_t, Eigen::Dynamic, Eigen::Dynamic>
+compute_ordering_triangle(const mesh::Mesh& mesh)
+{
+  const int num_cells = mesh.num_entities(mesh.topology().dim());
+  const int num_permutations = get_num_permutations(mesh.cell_type());
+  Eigen::Array<std::int8_t, Eigen::Dynamic, Eigen::Dynamic> cell_orders(
+      num_cells, num_permutations);
+
+  // Set orders for each cell
+  for (int cell_n = 0; cell_n < num_cells; ++cell_n)
+  {
+    const mesh::MeshEntity cell(mesh, 2, cell_n);
+    const std::int32_t* vertices = cell.entities(0);
+
+    // Set the orders for the edge flips
+    cell_orders(cell_n, 0) = (vertices[1] > vertices[2]);
+    cell_orders(cell_n, 1) = (vertices[0] > vertices[2]);
+    cell_orders(cell_n, 2) = (vertices[0] > vertices[1]);
+
+    // Set the orders for the face rotation and reflection
+    const std::array<int, 2> tri_orders
+        = calculate_triangle_orders(vertices[0], vertices[1], vertices[2]);
+    cell_orders(cell_n, 3) = tri_orders[0];
+    cell_orders(cell_n, 4) = tri_orders[0];
+  }
+
+  return cell_orders;
+}
+//-----------------------------------------------------------------------------
+Eigen::Array<std::int8_t, Eigen::Dynamic, Eigen::Dynamic>
+compute_ordering_tetrahedron(const mesh::Mesh& mesh)
+{
+  const int num_cells = mesh.num_entities(mesh.topology().dim());
+  const int num_permutations = get_num_permutations(mesh.cell_type());
+  Eigen::Array<std::int8_t, Eigen::Dynamic, Eigen::Dynamic> cell_orders(
+      num_cells, num_permutations);
+
+  // Set orders for each cell
+  for (int cell_n = 0; cell_n < num_cells; ++cell_n)
+  {
+    const mesh::MeshEntity cell(mesh, 3, cell_n);
+    const std::int32_t* vertices = cell.entities(0);
+
+    // Set the orders for the edge flips
+    cell_orders(cell_n, 0) = (vertices[2] > vertices[3]);
+    cell_orders(cell_n, 1) = (vertices[1] > vertices[3]);
+    cell_orders(cell_n, 2) = (vertices[1] > vertices[2]);
+    cell_orders(cell_n, 3) = (vertices[0] > vertices[3]);
+    cell_orders(cell_n, 4) = (vertices[0] > vertices[2]);
+    cell_orders(cell_n, 5) = (vertices[0] > vertices[1]);
+
+    // Set the orders for the face rotations and reflections
+    const std::array<int, 2> tri_orders0
+        = calculate_triangle_orders(vertices[1], vertices[2], vertices[3]);
+    cell_orders(cell_n, 6) = tri_orders0[0];
+    cell_orders(cell_n, 7) = tri_orders0[1];
+    const std::array<int, 2> tri_orders1
```

Comment:
Do these need to be a new array every time `tri_orders0`, `tri_orders1` etc. Maybe better to just reuse?

Reply:
I made that change. I felt that a static allocation for each was clearer than re-assigning.